Use endpoints informer for the endpoint controller

jsravn opened the pull request:
This substantially reduces the number of API calls made by the endpoint
controller. Currently the controller makes an API call per endpoint for
each service that is synced. When the 30s resync is triggered, this
results in an API call for every single endpoint in the cluster. This
quickly exceeds the default qps/burst limit of 20/30 even in small
clusters, leading to delays in endpoint updates.

This change modifies the controller to use the endpoint informer cache
for all endpoint GETs. This means we only make API calls for changes in
endpoints. As a result, qps only depends on the pod activity in the
cluster, rather than the number of services.

<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
-->

**What this PR does / why we need it**:

Address endpoint update delays as described in https://github.com/kubernetes/kubernetes/issues/47597.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

https://github.com/kubernetes/kubernetes/issues/47597

**Special notes for your reviewer**:

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
```

Commit:
Use endpoints informer for the endpoint controller

This substantially reduces the number of API calls made by the endpoint
controller. Currently the controller makes an API call per endpoint for
each service that is synced. When the 30s resync is triggered, this
results in an API call for every single endpoint in the cluster. This
quickly exceeds the default qps/burst limit of 20/30 even in small
clusters, leading to delays in endpoint updates.

This change modifies the controller to use the endpoint informer cache
for all endpoint GETs. This means we only make API calls for changes in
endpoints. As a result, qps only depends on the pod activity in the
cluster, rather than the number of services.

cmluciano commented:
@k8s-bot ok to test

shyamjvs commented on pkg/controller/endpoint/endpoints_controller.go:
```diff
@@ -119,6 +123,13 @@ type EndpointController struct {
 	// Added as a member to the struct to allow injection for testing.
 	podsSynced cache.InformerSynced
 
+	// endpointsLister is able to list/get pods and is populated by the shared informer passed to
```
nit: s/pods/endpoints/

smarterclayton commented on cmd/kube-controller-manager/app/core.go:
```diff
@@ -178,6 +178,7 @@ func startEndpointController(ctx ControllerContext) (bool, error) {
 	go endpointcontroller.NewEndpointController(
 		ctx.InformerFactory.Core().V1().Pods(),
 		ctx.InformerFactory.Core().V1().Services(),
+		ctx.InformerFactory.Core().V1().Endpoints(),
```
I didn't realize we weren't watching endpoints before - given that this controller is expected to know and mutate every endpoint, I think this is exactly what we should be doing.  It will have a memory impact, but will drop a significant chunk of QPS from the apiservers.

@wojtek-t @gmarek may change our kube-mark profiles, but I would expect it to be in a positive way.

gmarek replied:
I agree.

jsravn replied:
QPS drops substantially in my testing. I also don't see spurious updates I was seeing before (during a resync, about 20% of the endpoints would cause a PUT even though nothing changed - not sure why).

seh commented on pkg/controller/endpoint/endpoints_controller.go:
```diff
@@ -454,13 +465,14 @@ func (e *EndpointController) syncService(key string) error {
 // reboots).
 func (e *EndpointController) checkLeftoverEndpoints() {
-	list, err := e.client.Core().Endpoints(metav1.NamespaceAll).List(metav1.ListOptions{})
+	//list, err := e.client.Core().Endpoints(metav1.NamespaceAll).List(metav1.ListOptions{})
```
- [ ] Remove commented-out statement.

seh commented on pkg/controller/endpoint/endpoints_controller.go:
```diff
@@ -454,13 +465,14 @@ func (e *EndpointController) syncService(key string) error {
-	for i := range list.Items {
-		ep := &list.Items[i]
+	for i := range list {
+		ep := list[i]
```
- [ ] Why not use the two-value `range` clause here, given that [`EndpointsLister.List` returns a slice of pointers](https://github.com/kubernetes/client-go/blob/master/listers/core/v1/endpoints.go#L31)?  
```go
for _, ep := range list {
```

gmarek replied:
Yeah, copying a pointer is fine. In principle though we do use single-value `range` clauses to avoid copying large objects (which means +- all kubernetes API resources).

seh commented on pkg/controller/endpoint/endpoints_controller.go:
```diff
@@ -119,6 +123,13 @@ type EndpointController struct {
+	endpointsLister corelisters.EndpointsLister
+	// endpointsSynced returns true if the endpoints shared informer has been synced at least once.
+	// Added as a member to the struct to allow injection for testing.
+	endpointsSynced cache.InformerSynced
```
- [ ] Did you mean to include this in the set supplied to `controller.WaitForCacheSync` in `(*EndpointController).Run` at line 150?

smarterclayton replied:
Yes, please wait for endpoint cache to sync.  You don't want to make decisions on unintentionally blank data.

gmarek replied:
+1

jsravn replied:
Yup, I thought I had done it, will fix.

seh reviewed (changes requested):
It looks like we intended to hold up the controller's start-up procedure until the endpoints cache has synced, but I don't see the change that would make it so.

gmarek commented:
cc @shyamjvs

kargakis commented on pkg/controller/endpoint/endpoints_controller.go:
```diff
@@ -399,7 +410,7 @@ func (e *EndpointController) syncService(key string) error {
 
 	// See if there's actually an update here.
-	currentEndpoints, err := e.client.Core().Endpoints(service.Namespace).Get(service.Name, metav1.GetOptions{})
+	currentEndpoints, err := e.endpointsLister.Endpoints(service.Namespace).Get(service.Name)
```
You want to deep-copy these below before you mutate them.

gmarek replied:
+1. You need to do deep copy in line 432 (in case when you didn't enter `if` in line 415).

gmarek commented on pkg/controller/endpoint/endpoints_controller_test.go:
```diff
@@ -95,39 +95,43 @@ func makeTestServer(t *testing.T, namespace string, endpointsResponse serverResp
-	serviceStore cache.Store
+	podStore       cache.Store
+	serviceStore   cache.Store
+	endpointsStore cache.Store
```
Please add a test case that proves that we correctly wait for all informers to be synced.

gmarek commented on pkg/controller/endpoint/endpoints_controller_test.go:
```diff
@@ -170,23 +166,23 @@ func TestCheckLeftoverEndpoints(t *testing.T) {
 	}
 }
 
+//
```
remove

gmarek commented on pkg/controller/endpoint/endpoints_controller_test.go:
```diff
@@ -195,8 +191,10 @@ func TestSyncEndpointsProtocolTCP(t *testing.T) {
 			Ports:    []v1.ServicePort{{Port: 80, TargetPort: intstr.FromInt(8080), Protocol: "TCP"}},
 		},
 	})
+
```
remove

gmarek commented on pkg/controller/endpoint/endpoints_controller_test.go:
```diff
@@ -195,8 +191,10 @@ func TestSyncEndpointsProtocolTCP(t *testing.T) {
+
 	endpoints.syncService(ns + "/foo")
-	endpointsHandler.ValidateRequestCount(t, 2)
+
```
remove

gmarek commented on pkg/controller/endpoint/endpoints_controller_test.go:
```diff
@@ -235,8 +233,10 @@ func TestSyncEndpointsProtocolUDP(t *testing.T) {
 			Ports:    []v1.ServicePort{{Port: 80, TargetPort: intstr.FromInt(8080), Protocol: "UDP"}},
 		},
 	})
+
```
remove

gmarek commented on pkg/controller/endpoint/endpoints_controller_test.go:
```diff
@@ -235,8 +233,10 @@ func TestSyncEndpointsProtocolUDP(t *testing.T) {
+
 	endpoints.syncService(ns + "/foo")
-	endpointsHandler.ValidateRequestCount(t, 2)
+
```
remove

gmarek commented on pkg/controller/endpoint/endpoints_controller_test.go:
```diff
@@ -272,7 +272,9 @@ func TestSyncEndpointsItemsEmptySelectorSelectsAll(t *testing.T) {
 			Ports:    []v1.ServicePort{{Port: 80, Protocol: "TCP", TargetPort: intstr.FromInt(8080)}},
 		},
 	})
+
```
remove

gmarek commented on pkg/controller/endpoint/endpoints_controller_test.go:
```diff
@@ -272,7 +272,9 @@ func TestSyncEndpointsItemsEmptySelectorSelectsAll(t *testing.T) {
 	})
+
 	endpoints.syncService(ns + "/foo")
+
```
remove

gmarek commented on pkg/controller/endpoint/endpoints_controller_test.go:
```diff
@@ -308,7 +310,9 @@ func TestSyncEndpointsItemsEmptySelectorSelectsAllNotReady(t *testing.T) {
 			Ports:    []v1.ServicePort{{Port: 80, Protocol: "TCP", TargetPort: intstr.FromInt(8080)}},
 		},
 	})
+
```
remove

gmarek commented on pkg/controller/endpoint/endpoints_controller_test.go:
```diff
@@ -344,7 +348,9 @@ func TestSyncEndpointsItemsEmptySelectorSelectsAllMixed(t *testing.T) {
 			Ports:    []v1.ServicePort{{Port: 80, Protocol: "TCP", TargetPort: intstr.FromInt(8080)}},
 		},
 	})
+
```
remove

gmarek commented on pkg/controller/endpoint/endpoints_controller_test.go:
```diff
@@ -384,7 +390,9 @@ func TestSyncEndpointsItemsPreexisting(t *testing.T) {
 			Ports:    []v1.ServicePort{{Port: 80, Protocol: "TCP", TargetPort: intstr.FromInt(8080)}},
 		},
 	})
+
```
remove

gmarek commented on pkg/controller/endpoint/endpoints_controller_test.go:
```diff
@@ -423,14 +431,14 @@ func TestSyncEndpointsItemsPreexistingIdentical(t *testing.T) {
 			Ports:    []v1.ServicePort{{Port: 80, Protocol: "TCP", TargetPort: intstr.FromInt(8080)}},
 		},
 	})
+
```
remove

gmarek commented on pkg/controller/endpoint/endpoints_controller_test.go:
```diff
@@ -445,7 +453,9 @@ func TestSyncEndpointsItems(t *testing.T) {
 			},
 		},
 	})
+
```
remove

gmarek commented on pkg/controller/endpoint/endpoints_controller_test.go:
```diff
@@ -490,7 +499,9 @@ func TestSyncEndpointsItemsWithLabels(t *testing.T) {
 			},
 		},
 	})
+
```
remove

gmarek commented on pkg/controller/endpoint/endpoints_controller_test.go:
```diff
@@ -546,7 +557,9 @@ func TestSyncEndpointsItemsPreexistingLabelsChange(t *testing.T) {
 			Ports:    []v1.ServicePort{{Port: 80, Protocol: "TCP", TargetPort: intstr.FromInt(8080)}},
 		},
 	})
+
```
remove

gmarek commented:
Few nits, two comments, but generally we waited too long with a PR such as this. I'm inclined to put it into 1.7

jsravn commented:
I addressed the review comments. I wasn't too sure on the best way to test waiting for the caches, and this doesn't seem to be tested in the other controllers that I looked at. #Run isn't easily testable, so I extracted the wait into a separate method I could test.

gmarek commented:
Actually I though about testing it through 'Run'. E.g. populating two informers (pod and service), not syncing the last one (endpoint). Wait some time (I know it's bad, but we don't have better way to do this + this shouldn't cause any flakes) and make sure that nothing was done by Workers (e.g. no Endpoint object was created). Then sync last informer and check that Endpoint is eventually created. WDYT?

jsravn commented:
@gmarek I started going down that route, but was worried about the potential for flakiness. The test was also more complex. I'd probably need to make the worker loop period configurable since 1s is too long I think for a test to wait. If you're happy with that then I'll go ahead and do it.

gmarek commented:
I think it should be fine. Flakyness in one direction shouldn't happen (and if it does it actually means that we don't wait for sync), and for the other we may have short polling loop (e.g. 500ms) with long timeout (e.g. 1m), so that the test will generally finish quickly but allow a lot of slack if something takes long time.

jsravn commented:
@gmarek Okay, I made that change, let me know what you think. I'll be around later to follow up and hopefully get this PR finished.

gmarek commented:
/test pull-kubernetes-e2e-gce-etcd3

gmarek commented:
/lgtm

jsravn commented:
/assign @deads2k

deads2k commented:
/approve

jsravn commented:
Is the plan to get this in 1.7? Also, would it be possible to cherry pick to 1.6? Thanks!

smarterclayton commented:
It's too risky and too late for 1.7.  It might be a backport but it's also
risky to change this code with a lot of testing

On Jun 22, 2017, at 3:59 PM, James Ravn <notifications@github.com> wrote:

Is the plan to get this in 1.7? Also, would it be possible to cherry pick
to 1.6? Thanks!

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/47731#issuecomment-310500881>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/ABG_p_aW_gdzdiGc4VFipnjt0_AGcwEOks5sGtWlgaJpZM4N-b4j>
.

smarterclayton commented:
*Without*

smarterclayton commented:
Build is failing

jsravn commented:
I rebased on master to see what was broken (was an import, client-go/api package moved in master). Fixed it now, didn't mean to reset the lgtm.

kargakis commented on pkg/controller/endpoint/endpoints_controller_test.go:
```diff
@@ -35,9 +36,11 @@ import (
 	"k8s.io/kubernetes/pkg/client/clientset_generated/clientset"
 	informers "k8s.io/kubernetes/pkg/client/informers/informers_generated/externalversions"
 	"k8s.io/kubernetes/pkg/controller"
+	"time"
```
Can you move this import with the rest of the stdlib imports?

kargakis commented on pkg/controller/endpoint/endpoints_controller.go:
```diff
@@ -41,6 +41,7 @@ import (
 	"k8s.io/kubernetes/pkg/util/metrics"
 
 	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/api"
```
Move this with the rest of the k8s imports

jsravn commented:
@kargakis Done

kargakis commented:
@jsravn thanks, please squash and I'll retag with lgtm

jsravn commented:
@kargakis Done

kargakis commented:
/lgtm

jsravn commented:
FYI: After backporting this to our 1.6 cluster, it has fixed all issues we were having with endpoints.

gmarek commented:
@dchen1107 @smarterclayton - should we consider putting it in 1.7.X release?

wojtek-t commented:
> @dchen1107 @smarterclayton - should we consider putting it in 1.7.X release?

+1

wojtek-t commented:
I'm happy to create a cherrypick PR.

jsravn commented:
@wojtek-t Did you cherrypick to 1.7?

jsravn commented:
It's funny how the delayed ep updates subtly break things. We accidentally updated to 1.7, and I forgot we hadn't cherry picked this fix. As a result we've had increases in DNS lookup failures (such as when kube-dns is redeployed/moves nodes) and increases in upstream error rates. I'm going to cherry pick this into 1.7, even if a bit late...
